components/p3a/nitro_utils/attestation.cc

```diff
@@ -0,0 +1,264 @@
+#include "brave/components/p3a/nitro_utils/cose.h"
+#include "crypto/random.h"
+#include "net/base/url_util.h"
+#include "net/cert/pki/parsed_certificate.h"
+#include "net/cert/x509_certificate.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "services/network/url_loader_factory.h"
+#include "third_party/boringssl/src/include/openssl/pool.h"
+
+namespace nitro_utils {
+
+namespace {
+
+constexpr size_t kAttestationBodyMaxSize = 16384;
+constexpr net::SHA256HashValue kAWSRootCertFP{
+    .data = {0x64, 0x1A, 0x03, 0x21, 0xA3, 0xE2, 0x44, 0xEF, 0xE4, 0x56, 0x46,
+             0x31, 0x95, 0xD6, 0x06, 0x31, 0x7E, 0xD7, 0xCD, 0xCC, 0x3C, 0x17,
+             0x56, 0xE0, 0x98, 0x93, 0xF3, 0xC6, 0x8F, 0x79, 0xBB, 0x5B}};
+
+net::NetworkTrafficAnnotationTag AttestationAnnotation() {
+  return net::DefineNetworkTrafficAnnotation("nitro_utils_attestation", R"(
+    semantics {
+      sender:
+        "AWS Nitro Enclave Attestation/Validation"
+      description:
+        "Used to validate an AWS Nitro Enclave attestation document. "
+        "Nitro Enclaves are used in services such as P3A, to protect user anonymity."
+      trigger:
+        "Attestation requests are automatically sent at intervals while Brave "
+        "is running."
+      data: "A random nonce, and attestation document"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "This feature can be disabled via the P3A setting."
+      policy_exception_justification:
+        "Not implemented."
+    })");
+}
+
+bool VerifyNonce(const cbor::Value::MapValue& cose_map,
+                 const std::vector<uint8_t>& orig_nonce) {
+  cbor::Value::MapValue::const_iterator nonce_it =
+      cose_map.find(cbor::Value("nonce"));
+  if (nonce_it == cose_map.end() || !nonce_it->second.is_bytestring()) {
+    LOG(ERROR) << "Nitro verification: nonce is missing or is not bstr";
+    return false;
+  }
+
+  if (nonce_it->second.GetBytestring() != orig_nonce) {
+    LOG(ERROR) << "Nitro verification: nonce mismatch";
+    return false;
+  }
+  return true;
+}
+
+bool VerifyUserDataKey(scoped_refptr<net::X509Certificate> server_cert,
+                       const cbor::Value::MapValue& cose_map) {
+  cbor::Value::MapValue::const_iterator user_data_it =
+      cose_map.find(cbor::Value("user_data"));
+  if (user_data_it == cose_map.end() || !user_data_it->second.is_bytestring() ||
+      user_data_it->second.GetBytestring().size() != 32) {
+    LOG(ERROR) << "Nitro verification: user data is missing or is not bstr "
+               << "or is not 32 bytes";
+    return false;
+  }
+  net::SHA256HashValue server_cert_fp =
+      net::X509Certificate::CalculateFingerprint256(server_cert->cert_buffer());
```

Comment:
i guess we need to somehow check that the cert and the whole response are valid, otherwise it looks like using pure data from network. Starting with non-null server_cert and is_valid()  SSLInfo

Reply:
Found out that we do this prior to this call, so in this function we can thrown in additional DCHECK's